### PR TITLE
[FEAT] Restock by Shop

### PR DESCRIPTION
**Pull request by eliasSFL**
# Description

This PR enhances the restock system. 

Restock Options
image

Restock Options (with gem experiment)
image

Confirmation screen
image

Fixes #issue

# What needs to be tested by the reviewer?

Run FE and BE

1. Clear the `stock` field from your gamestate
- Go to each shop and restock your stock. 
- They should only restock the stock for the store you're currently in

2. Buy a partial amount of stock
- Go to each shop to restock your stock.
- They should only restock the stock for the store you're currently in, but doesn't remove stock of other items

3. Place down Warehouse and Toolshed
- Go to each shop and restock your stock. 
- They should only restock the stock for the store you're currently in, to the boosted amount

Additional tests:
Test to see UI for `GEM_BOOSTS` experiment, make sure those text only appear when they have the experiment and not when they don't

# Checklist:

- [ ] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [ ] Screenshot if it includes any UI changes
- [ ] I have read the contributing guidelines and agree to the T&Cs
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/features/game/events/landExpansion/enhancedRestock.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  action: EnhancedRestockAction;
+};
+
+type RestockObject = {
+  restockItem: object;
+  gemPrice: number;
+  shopName: "market" | "workbench" | "treasure shop";
+};
+
+export const RestockItems: Record<RestockNPC, RestockObject> = {
```
Comment (brian-man) [inline src/features/game/events/landExpansion/enhancedRestock.ts:29]: In aggregate these are more expensive than the current single restock.

Further, betty + blacksmith are also more expensive than today.

So, just on that alone it's a downgrade. :|

Perhaps some tuning to make this more appealing? @adamhannigan

**Comment by Dionis404** [comment]
@eliasSFL 
Thank you so much for implementing this!

File: src/features/island/buildings/components/building/market/restock/EnhancedRestockModal.tsx [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  const { openModal } = useContext(ModalContext);
+
+  const { gameService, showAnimations } = useContext(Context);
+  const [
+    {
+      context: { state },
+    },
+  ] = useActor(gameService);
+
+  const canRestock = state.inventory["Gem"]?.gte(20);
```
Comment (eliasSFL) [inline src/features/island/buildings/components/building/market/restock/EnhancedRestockModal.tsx:50]:
```suggestion
  const canRestock = state.inventory["Gem"]?.gte(gemPrice);
```

File: src/features/island/buildings/components/building/market/restock/EnhancedRestockModal.tsx [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    if (!remainingStock) {
+      // return total stock amount
+      return amount;
+    } else {
+      // else return difference between total and remaining stock
+      return amount.sub(remainingStock);
+    }
+  };
+
+  const { gemPrice, shopName, restockItem, categoryLabel } = RestockItems[npc];
```
Comment (eliasSFL) [inline src/features/island/buildings/components/building/market/restock/EnhancedRestockModal.tsx:86]:
```suggestion
```

File: src/lib/i18n/dictionaries/dictionary.json [hunk 4]
```diff
@@ -3740,7 +3738,21 @@
   "crafting.readySoon": "{{name}} will be ready soon...",
   "crafting.expansionSoon": "Your land will be ready soon...",
   "crafting.viewRecipes": "View recipes",
-  "gems.buyReplenish": "Would you like to replenish the shops now for 20 Gems?",
+  "gems.buyReplenish": "Would you like to restock all the shops for 20 Gems?",
+  "gems.buyReplenish.enhanced": "Would you like to restock the {{shopName}} for {{gemPrice}} Gems?",
+  "restock.shop": "{{shopName}} Restock",
+  "restock.full": "Full Restock",
+  "restock.outOfStock": "Looks like you ran out of stock from {{npc}}!",
```
Comment (adamhannigan) [inline src/lib/i18n/dictionaries/dictionary.json:3745]: I think we can shorten this + make it more direct: `Betty ran out of stock. Choose your desired restock option below:`

**Review by adamhannigan — DISMISSED** [review]
LGTM - just a suggestion for the text to show 👍

**Review by adamhannigan — APPROVED** [review]

**Pull request merged**